gapis/resolve/dependencygraph/footprint.go

```diff
@@ -0,0 +1,224 @@
+// which belongs to the command or subcomand indexed by the given SubCmdIdx. In
+// case the SubCmdIdx is invalid or a valid Behaviour index is not found, error
+// will be logged and uint64(0) will be returned.
+func (fpt *Footprint) GetBehaviourIndex(ctx context.Context,
+	fci api.SubCmdIdx) uint64 {
+	v := fpt.commandIndexToBehaviourIndexLookupTable.Value(fci)
+	if v != nil {
+		if u, ok := v.(uint64); ok {
+			return u
+		}
+		log.E(ctx, "Invalid behaviour index: %v is not a uint64", v)
+		return uint64(0)
+	}
+	log.E(ctx, "Cannot get behaviour index for command indexed with: %v", fci)
+	return uint64(0)
+}
+
+// AddBehaviour adds the given Behaviour to the Footprint and updates the
+// internal mapping from SubCmdIdx to the last Behaviour that belongs to that
+// command or subcommand.
+func (fpt *Footprint) AddBehaviour(ctx context.Context, b *Behaviour) bool {
+	bi := uint64(len(fpt.Behaviours))
+	fci := b.BelongTo
+	fpt.commandIndexToBehaviourIndexLookupTable.SetValue(fci, bi)
+	fpt.Behaviours = append(fpt.Behaviours, b)
+	fpt.BehaviourIndices[b] = bi
+	return true
+}
+
+// FootprintBuilderProvider provides FootprintBuilder
+type FootprintBuilderProvider interface {
+	FootprintBuilder(context.Context) FootprintBuilder
+}
+
+// FootprintBuilder incrementally builds Footprint one command by one command.
+type FootprintBuilder interface {
+	BuildFootprint(context.Context, *api.State, *Footprint, api.CmdID, api.Cmd)
+}
+
+// GetFootprint returns a pointer to the resolved Footprint.
+func GetFootprint(ctx context.Context) (*Footprint, error) {
+	r, err := database.Build(ctx, &FootprintResolvable{Capture: capture.Get(ctx)})
+	if err != nil {
+		return nil, fmt.Errorf("Counld not get execution foot print: %v", err)
+	}
+	return r.(*Footprint), nil
+}
+
+// Resolve implements the database.Resolver interface.
+func (r *FootprintResolvable) Resolve(ctx context.Context) (interface{},
+	error) {
+	c, err := capture.ResolveFromPath(ctx, r.Capture)
+	if err != nil {
+		return nil, err
+	}
+	cmds := c.Commands
+	builders := map[api.API]FootprintBuilder{}
+
+	ft := NewFootprint(ctx, cmds)
+
+	s := c.NewState()
+	t0 := footprintBuildCounter.Start()
+	for i, cmd := range cmds {
+		a := cmd.API()
+		if _, ok := builders[a]; !ok {
+			if bp, ok := a.(FootprintBuilderProvider); ok {
+				builders[a] = bp.FootprintBuilder(ctx)
+			} else {
+				// API does not provide execution footprint info, always keep commands
+				// fromt such APIs alive.
+				bh := NewBehaviour(api.SubCmdIdx{uint64(i)}, &dummyMachine{})
+				bh.Alive = true
+				// Even if the command does not belong to an API that provides
+				// execution footprint info, we still need to mutate it in the new
+				// state, because following commands in other APIs may depends on the
+				// side effect of the this command.
+				if err := cmd.Mutate(ctx, s, nil); err != nil {
+					log.W(ctx, "Command %v %v: %v", api.CmdID(i), cmd, err)
```

Comment:
Looks like this could spam for traces that have issues. The report view should be the place to display these warnings to the user, not the log.